keps/sig-cli/2906-kustomize-function-catalog/README.md

```diff
@@ -0,0 +1,667 @@
+
+### Goals
+
+Develop an API (`kind`) for Kustomize that is focused on plugin provider discovery, as well as guidelines and recommendations for distribution of plugin providers and associated resources, such as schema definitions.
+
+A successful implementation of this API should have the following characteristics:
+
+1. Plugin based workflows are driven by seamless invocation of sets of plugin providers without individual out-of-band discovery or installation steps for specific providers.
+1. The new API is integrated with the existing Kustomize tool (i.e. `kustomize build`) through references provided in Kustomization, Component, and Composition resources.
+1. Eligible Kustomize functionality could be extracted and distributed as official extensions. This won't be completed as part of this KEP but the required changes to support this will be implemented. 
+
+### Non-Goals
+
+1. Support anything other than KRM-style plugins that follow the [functions spec](https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/docs/api-conventions/functions-spec.md)
+1. Directly implement capabilities to publish plugin providers or other resources to OCI registries 
+
+## Proposal
+
+In order to standardize plugin discovery, introduce a `Catalog` API `kind` recognized by `kustomize build`. One or more `Catalog` resources can be referenced from any Kustomize kind as either a local file or remote reference available as from an HTTP(s) endpoint or as an [OCI artifact](https://github.com/opencontainers/artifacts).
+
+A `Catalog` will contain a collection of one or more modules that can be used with a Kustomize resource. 
+
+A minimal example is shown below:
+
+```yaml
+apiVersion: kustomize.io/v1
+kind: Catalog
+metadata: 
+  name: "example-co-plugins"
+spec: 
+  modules: 
+  - apiVersion: example.com/v1
+    kind: JavaApplication
+    description: "A Kustomize plugin provider that can handle Java apps"
+    provider: 
+      container: 
+        image: example/module_providers/java:v1.0.0
+```
+
+This will enable a users to use plugin providers with provider information omitted, such as:
+
+```yaml
+# app/kustomization.yaml
+kind: Kustomization
+catalogs:
+  - https://example.com/plugins/catalog.yaml
+resources:
+- javaapp.yaml
+```
+
+```yaml
+# app/javaapp.yaml
+apiVersion: example.com/v1
+kind: JavaApplication
+spec:
+  application: team/my-app
+  version: v1.0
+```
+
+When this Kustomization is processed by `kustomize build`, the referenced catalog (or catalogs) will be used to locate a plugin provider that supports the apiVersion `example.com/v1` and kind `JavaApplication`. If found in one of the referenced catalogs, kustomize can determine the provider configuration without the need for the user to specify it in the kustomization resources directly. The catalogs will be searched in order specified. If more than one catalog defines the target apiVersion and kind, the first will be selected. 
```

Comment:
> In what scenarios would a user have multiple catalogs that define the same apiVersion and kind? Does selecting the first make sense for all such cases?

It's pure speculation at this point, but I could imagine a scenario where multiple private catalogs include some of the same basic plugins, and I'd say the strategy of picking the first match is one reasonable approach to accommodating that. If the end user actually wants to pick one or the other, they can always use an override in their Kustomization. I don't feel strongly though; we could also disallow multiple matches and be more flexible later if we get feedback saying it's needed.

> As an example discussed in meeting, we have several KRM functions that use ConfigMap as the function specification kind.

The use of `ConfigMap` as a functionConfig presents a different question. I see that some of the KPT functions support both; for those, the easy solution would be to include the fully specified type in the catalog. For the others, do those functions check for the ConfigMap GVK, or just the `data:` structure? If the latter, the Catalog could be written with a "fake" GVK that in practice the function itself ignores. That would allow the function to be trusted and invoked in Kustomize via Catalog, i.e. with:

```yaml
transformers:
- kind: SetLabels
  apiVersion: kpt.dev/v1
  data:
    foo: bar
```

However, for either orchestrator to be able to use a Catalog to drive imperative discovery or configuration authoring flows (which there seemed to be interest in at the meeting), a fully schematized KRM kind would be required.* Our original draft of this KEP included a field for associated schemas, but we removed it before publication because we concluded it is not necessary to the goals of this KEP as stated. TL;DR it would look a lot like a CRD and would have analogous potential applications. Do you think we should add that back in? If we do, I think it should be a quick note on this PR and then a follow-up can be made with details to allow for more focussed review.

\* AFAICT it would not work to use the ConfigMap kind, because the key-value pairs the functions expect are not really arbitrary and opaque, as a ConfigMap data is from the POV of the ConfigMap controller. They seem to be real fields recognized by the function that happen to fit into `map[string]string`; this is particularly clear in the case of kubeval, where the keys names are individually significant.

Reply:
> AFAICT it would not work to use the ConfigMap kind, because the key-value pairs the functions expect are not really arbitrary and opaque, as a ConfigMap data is from the POV of the ConfigMap controller. They seem to be real fields recognized by the function that happen to fit into map[string]string; this is particularly clear in the case of kubeval, where the keys names are individually significant.

This is true for functions like `kubeval`. But not true for functions like [`set-annotations`](https://catalog.kpt.dev/set-annotations/v0.1/) and [`set-labels`](https://catalog.kpt.dev/set-labels/v0.1/). For example, `set-labels` will use the arbitrary string key-value pairs from the `configMap` and add them to `metadata.labels`.

The KRM function spec doesn't require the `functionConfig` to have unique GVK. So it's a valid use case to have multiple different functions to use `ConfigMap` as `functionConfig`. We should support it.

Reply:
A seemingly simple solution to this would be to add an optional `name` field to the catalog.

For example:

catalog.yaml:
```
apiVersion: kustomize.io/v1
kind: Catalog
metadata: 
  name: "example-co-plugins"
spec: 
  modules: 
  - apiVersion:v1
    kind: ConfigMap
    name: kubeval-config # could be optional
    description: "kubeval function"
    provider: 
      container: 
        image: kubeval:v1.0.0
```

kustomization.yaml:
```
catalogs:
  - catalog.yaml
transformers:
  - kubeval-config.yaml
```

kubeval-config.yaml
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: kubeval-config
data: 
...
```

In this case, kustomize could use the entire GVKN of the transformer config resource to do the catalog lookup, allowing more flexibility of functions with duplicate GVKs. If the catalog entry has no name field, then it can just match the GVK. 

WDYT @KnVerey @jeremyrickard?

Reply:
I believe we discussed this in the collaboration meeting, and we decided to not support the ConfigMap use case. Ref: https://github.com/kubernetes/enhancements/pull/2954#discussion_r713347073

Reply:
Yes, we can mark this conversation resolved.